### PR TITLE
Add support for CBDT and COLR fonts

### PR DESCRIPTION
**Pull request by nulano**
Proper fix for #891, partial support for #3346

Changes proposed in this pull request:

 * properly detect glyph bitmap mode instead of guessing
 * use `FT_Convert_Bitmap` instead of a custom solution
 * add support for Google CBDT (if FreeType is compiled with libpng) and Microsoft COLR fonts

Apple SBIX and W3C SVG fonts are not yet supported by FreeType.

Example:

```python
from PIL import Image, ImageDraw, ImageFont

def test(font, out_name):
    fnt = ImageFont.truetype(font, size=109, layout_engine=ImageFont.LAYOUT_RAQM)
    im = Image.new("RGBA", (600, 600), (100, 100, 100, 100))
    draw = ImageDraw.Draw(im)
    draw.text((0, 32), "a\u263A", fill="#faa2", embedded_color=True, font=fnt)
    draw.text((0, 132), "a\u263A", fill="#faa8", embedded_color=True, font=fnt)
    draw.text((0, 232), "a\u263A", fill="#faa", embedded_color=True, font=fnt)
    draw.text((0, 332), "\U0001F3AC\U0001F44B\U0001F3FB\U0001F44B\U0001F3FF", fill="white", embedded_color=True, font=fnt)
    draw.text((0, 432), "a\u263A", fill="#faa2", font=fnt)
    im.show()
    im.save(f"testemoji_{out_name}.png")

test(r"C:\Users\Nulano\Downloads\NotoColorEmoji.ttf", "cbdt")
test("seguiemj.ttf", "colr")
```

NotoColorEmoji.ttf (CBDT):
testemoji_cbdt

seguiemj.ttf (COLR):
testemoji_colr

### CODE REVIEW

File: Tests/test_imagefont_bitmap.py [hunk 1]
```diff
@@ -39,4 +40,22 @@ def test_similar():
         fill=(0, 0, 0),
         font=font_outline,
     )
-    assert_image_similar(im_bitmap, im_outline, 20)
+    assert_image_similar(im_bitmap, im_outline, 4)
+
+
+@skip_unless_feature("freetype2")
+@pytest.mark.parametrize("bpp", (1, 2, 4, 8))
+def test_bitmap_font(bpp):
```
Comment (nulano) [inline Tests/test_imagefont_bitmap.py:48]: This test is sufficient to replace the test above. Would it make sense to remove `test_imagefont_bitmap.test_similar` and move this new test to `test_imagefont`?

**Comment by nulano** [comment]
Big-endian s390 is failing, but I'm not sure I can fix it without seeing the output image.

**Comment by hugovk** [comment]
This increases the size of the sdist by about 20% from 58M to 69M.

Are all these fonts necessary, or can we still be confident in testing with existing fonts (I guess they're not CBDT or COLR), or just one of the smaller additions?

**Comment by nulano** [comment]
> This increases the size of the sdist by about 20% from 58M to 69M.
> 
> Are all these fonts necessary, or can we still be confident in testing with existing fonts (I guess they're not CBDT or COLR), or just one of the smaller additions?

`DejaVuSans-24-*.ttf` (6MB total) can be replaced by a different font or a cut-down version with just a few glyphs for testing.

The biggest chunk is the CBDT font `NotoColorEmoji.ttf`; CBDT fonts are essentially just a collection of bitmaps, so any font will be large. If the license allows it, stripping out all except the tested bitmaps might be possible.

`BungeeColor-Regular_colr_Windows.ttf` (COLR font) is just 73.6 KB.

**Comment by radarhere** [comment]
> Big-endian s390 is failing, but I'm not sure I can fix it without seeing the output image.

Here you go - generated with [this commit](https://github.com/radarhere/Pillow/commit/1acca668a33c51ad166da754b29db4fa017404e0) forked from your branch.
standard_embedded_fail

**Comment by nulano** [comment]
> > Big-endian s390 is failing, but I'm not sure I can fix it without seeing the output image.
> 
> Here you go - generated with [this commit](https://github.com/radarhere/Pillow/commit/1acca668a33c51ad166da754b29db4fa017404e0) forked from your branch.
> standard_embedded_fail

Ah, I had the right idea, but forgot `long long` is 8 bytes...

Thanks for the help.

**Comment by nulano** [comment]
> This increases the size of the sdist by about 20% from 58M to 69M.
> 
> Are all these fonts necessary, or can we still be confident in testing with existing fonts (I guess they're not CBDT or COLR), or just one of the smaller additions?

I've stripped down the DejaVuSans-24-*.ttf fonts to ASCII and used an older version of NotoColorEmoji.ttf (5.1MB). Is this sufficient?
I can't find any other CBDT fonts or a simple way to edit them.

**Comment by hugovk** [comment]
Looks like we also need the setuptools workaround for 3.8 (like https://github.com/python-pillow/Pillow/pull/4958) for Windows/GHA: 

```
RuntimeWarning: The _imaging extension was built for another version of Pillow or PIL:
Core version: None
Pillow version: 8.0.0.dev0
Traceback (most recent call last):
  File "selftest.py", line 6, in <module>
    from PIL import Image, features
  File "<frozen zipimport>", line 259, in load_module
  File "C:\hostedtoolcache\windows\Python\3.8.6\x86\lib\site-packages\pillow-8.0.0.dev0-py3.8-win32.egg\PIL\Image.py", line 97, in <module>
ImportError: The _imaging extension was built for another version of Pillow or PIL:
Core version: None
Pillow version: 8.0.0.dev0
Error: Process completed with exit code 1.
```

https://github.com/python-pillow/Pillow/pull/4955/checks?check_run_id=1239105264

File: Tests/test_imagefont.py [hunk 2]
```diff
@@ -840,6 +840,102 @@ def test_anchor_invalid(self):
...
+        txt = "Hello World!"
+        ttf = ImageFont.truetype(FONT_PATH, 40, layout_engine=self.LAYOUT_ENGINE)
+        ttf.getsize(txt)
+
+        img = Image.new("RGB", (300, 64), "white")
+        d = ImageDraw.Draw(img)
+        d.text((10, 10), txt, font=ttf, fill="#fa6", embedded_color=True)
+
+        with Image.open("Tests/images/standard_embedded.png") as expected:
+            assert_image_similar(img, expected, max(self.metrics["multiline"], 3))
```
Comment (hugovk) [inline Tests/test_imagefont.py:853]: nit
```suggestion
        im = Image.new("RGB", (300, 64), "white")
        d = ImageDraw.Draw(im)
        d.text((10, 10), txt, font=ttf, fill="#fa6", embedded_color=True)

        with Image.open("Tests/images/standard_embedded.png") as expected:
            assert_image_similar(im, expected, max(self.metrics["multiline"], 3))
```

File: Tests/test_imagefont.py [hunk 3]
```diff
@@ -840,6 +840,102 @@ def test_anchor_invalid(self):
...
+
+            im = Image.new("RGB", (150, 150), "white")
+            d = ImageDraw.Draw(im)
+
+            d.text((10, 10), "\U0001f469", embedded_color=True, font=font)
+
+            with Image.open("Tests/images/cbdt_notocoloremoji.png") as expected:
+                assert_image_similar(im, expected, self.metrics["multiline"])
+        except IOError as ex:
+            assert str(ex) in ("unimplemented feature", "unknown file format")
```
Comment (hugovk) [inline Tests/test_imagefont.py:874]:
```suggestion
        except IOError as e:
            assert str(e) in ("unimplemented feature", "unknown file format")
```

File: Tests/test_imagefont.py [hunk 4]
```diff
@@ -840,6 +840,102 @@ def test_anchor_invalid(self):
...
+
+            im = Image.new("RGB", (150, 150), "white")
+            d = ImageDraw.Draw(im)
+
+            d.text((10, 10), "\U0001f469", "black", font=font)
+
+            with Image.open("Tests/images/cbdt_notocoloremoji_mask.png") as expected:
+                assert_image_similar(im, expected, self.metrics["multiline"])
+        except IOError as ex:
+            assert str(ex) in ("unimplemented feature", "unknown file format")
```
Comment (hugovk) [inline Tests/test_imagefont.py:896]:
```suggestion
        except IOError as e:
            assert str(e) in ("unimplemented feature", "unknown file format")
```

**Comment by nulano** [comment]
#4958 already includes 3.8, this branch just needs to be rebased or merge master (I think re-running the checks should suffice)

File: Tests/test_imagefont.py [hunk 5]
```diff
@@ -840,6 +840,102 @@ def test_anchor_invalid(self):
...
+            pytest.skip("freetype compiled without libpng or unsupported")
+
+    def test_colr(self):
+        if parse_version(features.version_module("freetype2")) < parse_version(
+            "2.10.0"
+        ):
+            pytest.skip("Freetype 2.10.0 or newer required")
```
Comment (hugovk) [inline Tests/test_imagefont.py:903]: There's quite a few Freetype version check and skips, let's put them into a helper, for example:

```python
def skip_unless_freetype(required_version, message):
```

File: Tests/test_imagefont_bitmap.py (hunk 1 above)
Comment (hugovk) [reply Tests/test_imagefont_bitmap.py:48]: Okay, let's do that.

File: docs/reference/ImageDraw.rst [hunk 6]
```diff
@@ -399,6 +404,19 @@ Methods
...
+                     .. versionadded:: 6.2.0
+
+    :param stroke_fill: Color to use for the text stroke. If not given, will default to
+                        the ``fill`` parameter.
+
+                     .. versionadded:: 6.2.0
+
+    :param embedded_color: Whether to use font embedded color glyphs (COLR or CBDT).
+
+                    .. versionadded:: 8.0.0
```
Comment (hugovk) [inline docs/reference/ImageDraw.rst:418]:
```suggestion
                     .. versionadded:: 8.0.0
```

**Review by hugovk — APPROVED** [review]

File: Tests/test_imagefont_bitmap.py (hunk 1 above)
Comment (nulano) [reply Tests/test_imagefont_bitmap.py]: Done, also added Raqm version of test.

**Comment by nulano** [comment]
Removing `test_imagefont_bitmap.test_similar` (in favor of the new test `test_imagefont.test_bitmap_font`) also allows removing 0.9MB `DejaVuSans-bitmap.ttf`.

File: Tests/test_imagefont.py (hunk 5 above)
Comment (nulano) [reply Tests/test_imagefont.py]: I went with the following to let the check be reusable for other features:
```python
def skip_unless_feature_version(feature, version_required, reason=None):
```

**Comment by hugovk** [comment]
Thank you! Please can you add release notes for this, plus for any of your other PRs since the last release?

**Comment by NightMachinary** [comment]
Apple font is fine on my system, FYI.
image

Is there a way to use one font for emojis and another for the normal text?

**Comment by nulano** [comment]
> Apple font is fine on my system, FYI.

Yep, figured that out a few weeks after this PR (#5155). 
 
> Is there a way to use one font for emojis and another for the normal text?

Not automatically, see #4808. If you are fine with splitting up the text yourself, you can use the approach outlined here: https://github.com/python-pillow/Pillow/issues/3346#issuecomment-683984319.

**Pull request merged**